web-animations-1/Overview.bs

```diff
@@ -4580,8 +4595,16 @@ dictionary OptionalEffectTiming {
     time taken to complete a single iteration of the <a>animation
     effect</a>.
 
-    In this level of this specification, the string value <code>auto</code>
-    is treated as the value zero for the purpose of timing model calculations
+    If the <a lt="timeline duration">duration</dfn> of the Animation's
+    <a>timeline</a> is a percentage, a non <code>auto</code> duration will
+    be treated as 0. Future levels of this specification are expected to
+    introduce support for scaling duration, iterations and start/end delays
+    to fill a percentage timeline duration.
```

Comment:
We probably should update the paragraph before this, and then describe the syntax for percentages (and how they are encoded as strings).

Also, it seems like the duration of a timeline can't be a percentage?

Reply:
Hmm, my thinking was that internally we would recognize that the timeline duration is a percentage even though the value by inspection would be 1.0 (consistent with getComputedStyle). However, I'm guessing the issue is that if it's not represented in the WebIDL return value we can't recognize this.

Some options:
* Define and use a subclass of AnimationTimeline (i.e. ProgressTimeline) and check the instance type to determine which behavior to apply.
* Use a string value for duration

Any other ideas? Any recommendations?

Reply:
Oh good point. I think it's probably helpful to ask what we want `getTiming()` and `getComputedTiming()` to return when we later introduce effects that are a percentage of their parent group's time (or ancestor timeline if we introduced fixed duration document timelines for example).

I think in that case we might want something like:

* `getTiming().duration` to return [`CSSNumberish`](https://drafts.css-houdini.org/css-typed-om-1/#typedefdef-cssnumberish), i.e. a `double` in the case of a fixed time (for compatibility) or a `CSSNumberValue` to represent a percentage.
* `getComputedTiming().duration` to return a number which is either the resolved duration in ms (in part for compatibility, and in part so we can determine the actual duration for complex group effects) or a value between [0, 1] or [0, 100] to represent a percentage of an auto duration timeline.

What do you think?

Reply:
I like this, if we do the same for the timeline duration then we can know when a defined duration on the timeline is a time (where time-based values in the animations don't need to be dropped) or a percentage (where we convert time based animations to their respective proportions). I'll work on making this change.

Reply:
Actually we already specify that the return value from getTiming() must be auto if the duration is specified as auto so we should be returning the string 'auto' from getTiming which will round trip. I still like the idea of using CSSNumberish for the timeline duration though so that we clearly know whether the duration is time-based or progress (percentage) based.

Reply:
Yes, sorry I was getting a bit confused because I forgot we are only trying to cover "auto" in this patch.

(My primary concern is making sure we have a clear path for supporting percentages in general in future so I am trying to think about how we will cover that.)

Reply:
We could consider returning CSSNumberish from computed timing - that would sidestep the issue of converting the percentages to a double value. It will mean changing a lot of values to be CSSNumberish, but this might be cleaner in the long run. WDYT?

Reply:
I'm not really sure. Once we are able to use percentages that resolve to times (e.g. `duration` is 50% of my parent group) I wonder if it would be useful to have `getComputedTiming()` return the actual number of seconds it resolves to? In which case, should percentages that resolve against scroll timelines resolve to pixels?

Unfortunately `getComputedStyle` doesn't give an entirely clear precedent here since it sometimes resolves percentages (e.g. for `width` and `height`) but otherwise doesn't. I guess the intention is that it should always return percentages. So maybe we should return CSSNumberish instead?

It would be interesting to know if anyone else has an opinion on this.

Reply:
> I'm not really sure. Once we are able to use percentages that resolve to times (e.g. `duration` is 50% of my parent group) I wonder if it would be useful to have `getComputedTiming()` return the actual number of seconds it resolves to?

I could see this being useful, and might be worth doing.

> In which case, should percentages that resolve against scroll timelines resolve to pixels?

This depends on what you consider to be the underlying value of the ScrollTimeline right? It could be pixels but it could also only operate in percentage of scroll or percentage of duration?

> Unfortunately `getComputedStyle` doesn't give an entirely clear precedent here since it sometimes resolves percentages (e.g. for `width` and `height`) but otherwise doesn't. I guess the intention is that it should always return percentages. So maybe we should return CSSNumberish instead?

I think returning CSSNumberish and keeping percentages in the computed timing makes sense. One thing that I believe supports this is this note from [css-cascade](https://drafts.csswg.org/css-cascade/#computed) (emphasis mine):

> Note: In general, the computed value resolves the specified value as far as possible **without laying out the document** or performing other expensive or hard-to-parallelize operations, such as resolving network requests or retrieving values other than from the element and its parent.

cc/ @andruud

Reply:
Leaving percentages intact sounds entirely reasonable.

Although depending on layout in this case is probably less painful than normal, given the snapshotting described in [Avoiding cycles with layout](https://drafts.csswg.org/scroll-animations-1/#avoiding-cycles), which means we can resolve the percentages against whatever is known from the previous snapshot?

> returning CSSNumberish

Note that (from brief inspection of css-typed-om & Blink): there seems to be an "understanding" that a _return value_ of `CSSNumberish` means you always get a `CSSNumericValue`. So _readonly_ `CSSNumberish` attributes would probably ideally be `CSSNumericValue` attributes. So since this would be the first deviation from this (I think), we should probably specify under which circumstances we can expect a raw double.

Reply:
> Note that (from brief inspection of css-typed-om & Blink): there seems to be an "understanding" that a _return value_ of `CSSNumberish` means you always get a `CSSNumericValue`. So _readonly_ `CSSNumberish` attributes would probably ideally be `CSSNumericValue` attributes. So since this would be the first deviation from this (I think), we should probably specify under which circumstances we can expect a raw double.

That sounds reasonable. If we were designing this from scratch I think we would always return a `CSSNumericValue` and it is only compatibility constraints that mean we sometimes need to return an actual number.